axum-extra/src/routing/typed.rs

```diff
@@ -295,3 +382,63 @@ impl_second_element_is!(T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13);
 impl_second_element_is!(T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14);
 impl_second_element_is!(T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15);
 impl_second_element_is!(T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16);
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    struct UsersShow {
+        id: i32,
+    }
+
+    // can't derive `TypedPath` because that references `::axum_extra` :(
+    impl fmt::Display for UsersShow {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(f, "/users/{}", self.id)
+        }
+    }
+
+    impl TypedPath for UsersShow {
+        const PATH: &'static str = "/users/:id";
+    }
+
+    #[derive(Serialize)]
+    struct Params {
+        foo: &'static str,
+        bar: i32,
+        baz: bool,
+    }
+
+    #[test]
+    fn with_params() {
+        let path = UsersShow { id: 1 }.with_query_params(Params {
+            foo: "foo",
+            bar: 123,
+            baz: true,
+        });
+
+        let uri = path.to_uri();
+
+        // according to [the spec] starting the params with `?&` is allowed specifically:
+        //
+        // > If bytes is the empty byte sequence, then continue.
+        //
+        // [the spec]: https://url.spec.whatwg.org/#urlencoded-parsing
+        assert_eq!(uri, "/users/1?&foo=foo&bar=123&baz=true");
+    }
+
+    #[test]
+    fn with_params_called_multiple_times() {
```

Comment:
We should also address how query params are merged? EG:
```rust
 let path = UsersShow { id: 1 }
            .with_query_params(Params {
                foo: "foo",
                bar: 123,
                baz: true,
            })
            .with_query_params([("foo", 1337)]); // does this override `foo` or get merged into a `foo[]="foo"&foo[]=1337`?
```

Reply:
Multiple values for the same query is valid, in fact that is what HTML forms do. Hence the existence of https://docs.rs/serde_html_form/latest/serde_html_form/index.html#

Reply:
🤘